oeps/oep-0012.rst

```diff
@@ -0,0 +1,520 @@
+Student View could ask each of its children to render itself as a fragment, and
+then could return a composite fragment containing all of them. This is
+essentially how XBlocks render their children today, and component views should
+work in the same way.
+
+The instructor dashboard feature in the `Django Component Views prototype`_ is
+another good example of how this can work. The dashboard renders itself, but
+when the user clicks to view a card, the dashboard renders the feature's own
+component.
+
+Should Django Component Views be side-effect free?
+--------------------------------------------------
+
+An interesting question is how Django component views should work within a
+post request. The simplest answer is to say that they work just the same way
+as regular Django views. However, there is the possibility of bad interactions
+if multiple Django component views are rendered into the same page within a
+post context, and they each perform updates.
+
+Should Django Component Views have a classification scheme?
+-----------------------------------------------------------
+
+If this proposed approach is successful, there could be a large number of
+registered component views. It would be useful to be able to classify views
+so that it is clear what there purpose is, e.g. this is a course outline view,
+this is an instructor dashboard view.
+
+For now, the approach taken has been to keep Django Component View as simple as
+regular Django Views, which don't have a classification scheme. Developers can
+use a different mechanism (probably Stevedore plugins) to declare the type of
+each view.
+
+For example, the Instructor Dashboard hackathon demo introduced a new
+``InstructorFeature`` plugin type, which in turn had a field for the
+component view used to render the feature. The Instructor Dashboard code
+could then ask Stevedore to return the list of all such features, in order
+to render them onto the page. For more details, see the implementation here:
+`Pluggable user interfaces hackathon PR`_.
+
+Backward Compatibility
+======================
+
+This proposed solution is all new functionality and so has no issues with
+backward compatibility.
+
+Future Directions
+=================
+
+Convert XBlocks to render using component views
+-----------------------------------------------
+
+There seem to be many benefits to XBlocks rendering themselves using component
+views, but the challenge is that today XBlocks explicitly do not make use of
+Django. A future rethinking of XBlocks based upon Django and utilizing component
+views would allow the following:
+
+* A runtime could allow the configuration of alternate UIs for certain blocks
+  that could be installed as plugins. For example, an Open edX operator could
+  choose to show a different UI for its "Course" tab. As another example,
+  a mobile runtime could show native mobile versions of certain blocks,
+  perhaps using `React Native`_.
```

Comment:
I think, in many cases, this ought to be a course-specific decision. Operator might have a default preference, but the right rendering is designed on how the course content ought to be consumed. Some courses are very linear, and some highly non-linear.